src/omv/imlib/draw.c

```diff
@@ -3072,6 +3072,36 @@ void imlib_draw_image(image_t *dst_img,
     // Force a deep copy if we are scaling.
     bool is_color_conversion_scaling = is_color_conversion && is_scaling;
 
+    #if (OMV_GPU_ENABLE == 1)
+    bool gpu_in_place_op = (dst_img->data == src_img->data)
+                           && (src_img_row_bytes >= dst_img_row_bytes) && (!is_color_conversion);
+
+    bool runnable_on_gpu = (rgb_channel < 0) && (!(hint & (IMAGE_HINT_AREA | IMAGE_HINT_BICUBIC))) && (!callback);
+
+    rectangle_t dst_rect;
+    dst_rect.x = dst_x_start;
+    dst_rect.y = dst_y_start;
+    dst_rect.w = dst_x_end - dst_rect.x;
+    dst_rect.h = dst_y_end - dst_rect.y;
+
+    rectangle_t src_rect;
+    src_rect.x = src_x_accum_reset >> 16;
+    src_rect.y = src_y_accum_reset >> 16;
+    src_rect.w = ((src_x_accum_reset + (src_x_frac * dst_rect.w)) >> 16) - src_rect.x;
+    src_rect.h = ((src_y_accum_reset + (src_y_frac * dst_rect.h)) >> 16) - src_rect.y;
+
+    image_hint_t gpu_hint = hint & (IMAGE_HINT_BILINEAR | IMAGE_HINT_BLACK_BACKGROUND);
+    gpu_hint |= (dst_delta_x < 0) ? IMAGE_HINT_HMIRROR : 0;
+    gpu_hint |= (dst_delta_y < 0) ? IMAGE_HINT_VFLIP : 0;
+
+    // Try to offload this to the GPU for processing before allocating a copy of the image.
+    if (gpu_in_place_op && (!is_jpeg) && (!is_png) && (!(hint & IMAGE_HINT_TRANSPOSE)) &&
+        runnable_on_gpu && (!omv_gpu_draw_image(src_img, &src_rect, dst_img, &dst_rect,
+                                                alpha, color_palette, alpha_palette, gpu_hint))) {
+        goto exit_cleanup;
+    }
+    #endif
+
```

Comment:
```suggestion
    #if (OMV_GPU_ENABLE == 1)
    bool scale_on_gpu = !callback &&
                        !is_color_conversion &&
                        !src_img->is_compressed &&
                        (rgb_channel < 0) &&
                        (src_img_row_bytes >= dst_img_row_bytes) &&
                        (dst_img->data == src_img->data) && // TODO why not scale on GPU if src!=dst
                        !(hint & (IMAGE_HINT_AREA | IMAGE_HINT_BICUBIC | IMAGE_HINT_TRANSPOSE));
    if (scale_on_gpu) {
        rectangle_t dst_rect = {
            .x = dst_x_start,
            .y = dst_y_start,
            .w = dst_x_end - dst_rect.x,
            .h = dst_y_end - dst_rect.y,
        };

        rectangle_t src_rect = {
            .x = src_x_accum_reset >> 16,
            .y = src_y_accum_reset >> 16,
            .w = ((src_x_accum_reset + (src_x_frac * dst_rect.w)) >> 16) - src_rect.x,
            .h = ((src_y_accum_reset + (src_y_frac * dst_rect.h)) >> 16) - src_rect.y,
        };

        image_hint_t gpu_hints = ((dst_delta_x < 0) ? IMAGE_HINT_HMIRROR : 0) |
                                 ((dst_delta_y < 0) ? IMAGE_HINT_VFLIP : 0) |
                                 (hint & (IMAGE_HINT_BILINEAR | IMAGE_HINT_BLACK_BACKGROUND));

        // Try to offload this to the GPU for processing before allocating a copy of the image.
        if (omv_gpu_draw_image(src_img, &src_rect, dst_img, &dst_rect,
                               alpha, color_palette, alpha_palette, gpu_hints)) {
            goto exit_cleanup;
        }
    }
    #endif

```

Reply:
I need the variables for the second gpu call which comes later in the code. The second call handles if src!=dst.

Reply:
Why do you need a second call? What's special about `src!=dst`? The GPU can scale either way.

Reply:
The second one handles if you are scaling after any type of color conversion or etc. which requires allocating a second image.

Yeah, I I guess though you can make the first one intercept the trivial out of place case. You don't need to wait till the second one to catch that.

Reply:
> if you are scaling after any type of color conversion or etc. which requires allocating a second image.

But scaling with color conversion is supported on both GPUs. Like?
```
img.to_grayscale(x_scale=0.5, y_scale=0.5, hint=image.BILINEAR)
```
No debayer though of course.

Reply:
Also if this is `(src_img_row_bytes >= dst_img_row_bytes)` to ensure that it's down-scaling because of that bug, it should either be moved to the GPU driver (if it can't handle it) or fixed.

Reply:
Here's a better explanation of what I was thinking when making this change. I was just adding in the functionality to cover making it work to handle the in-place case. The second call later on was still handling all other situations.

However, on pondering this more there's no reason the first call can't handle almost everything and then the second call just needs to handle scaling after yuv/bayer/jpeg/png conversion. Given this, I can actually reduce the complexity of the first call's to just checking `runnable_on_gpu && (!(hint & IMAGE_HINT_TRANSPOSE))` I think. The GPU call itself will return -1 if a jpeg/png/bayer/yuv image is passed.

> You should rename this is_color_conversion to something else like is_bayer_yuv_conversion.
Sure

> Some of the bool flags are only used one, please double check. For example do you really need is_color_conversion_scaling, need_deep_copy, is_jpeg, is_png ?
That's done to avoid a giant if statement filled with boolean logic.

Given me a bit, will update it...

Reply:
After debayer, JPEG/PNG decompression or YUV conversion it's a bit too late to benefit from GPU scaling speed up, isn't it? I'd just let the software fallback handle all other unsupported cases.

- Note that GPU can transpose while scaling, not tested and unsure if it can be done in place.
- Please remove the flags I mentioned.

Reply:
Done